examples_tests/15.LayeredCubeMapShadows/deriv_map_gen.comp

```diff
@@ -1,97 +1,46 @@
-	#if defined(GL_KHR_shader_subgroup_clustered) && defined(GL_KHR_shader_subgroup_basic)
-	
-	#define IRR_DDX_IS_DEFINED 1
-	// WARNING: No DDY here! Tpose and use DDX().
-
-	float DDX(in float x)
-	{
-		// The extension is supported but i'm getting "error: unknown builtin varying (named gl_SubgroupInvocationID) encountered" (?!?!?!?!?!)
-		int xormask = int((~gl_SubgroupInvocationID) << 31);
-		int fbits = floatBitsToInt(x);
-		// branchless sign swap of elements on even positions so that we get dx = (-[i])+[i+1] = [i+1]-[i]
-		return subgroupClusteredAdd(intBitsToFloat(fbits ^ xormask), 2);
-	}
-	#endif
-#endif
-#endif //#if 0 
-//possible use of extensions is turned off for now because untested due to:
-//1) my GTX660 doesn't support GL_NV_compute_shader_derivatives (actually seems like only high-end RTXs support this)
-//2) weird error "unknown builtin varying" concerning gl_SubgroupInvocationID even though GL_KHR_shader_subgroup_clustered is supported (driver bug?)
-
-// also distribution of work-groups on dimensions should depend on available GLSL extensions
-#if defined(IRR_DDX_IS_DEFINED) && defined(IRR_DDY_IS_DEFINED)
-layout (local_size_x = 16, local_size_y = 16) in;
-#else
-	#if XPASS
-	layout (local_size_x = 256) in;
-	#else
-	layout (local_size_y = 256) in;
-	#endif
-#endif
-
-#if !defined(IRR_DDX_IS_DEFINED) && !defined(IRR_DDY_IS_DEFINED)
-shared float smem[256]; //1KB
-#endif
-
-layout (binding = 7) uniform sampler2D bumpMapSampler;
-#if !defined(IRR_DDY_IS_DEFINED)
-layout (binding = 0, r16_snorm) uniform image2D derivativeMapImage;
-#else
-layout (binding = 0, rg16_snorm) uniform image2D derivativeMapImage;
-#endif
-
-layout (location = 0) uniform float uHeightScaleFactor = 0.1;
 
 void main()
 {
 	const ivec2 G_IDX = ivec2(gl_GlobalInvocationID.xy);
 	const ivec2 RESOLUTION = ivec2(gl_NumWorkGroups.xy * gl_WorkGroupSize.xy);
 	
-#if defined(IRR_DDX_IS_DEFINED) && defined(IRR_DDY_IS_DEFINED)
-	float smpl = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor;
-	float dx = DDX(smpl);
-	float dy = DDY(smpl);
-	
-	if ((G_IDX.x & 1) == 0 && (G_IDX.y & 1) == 0)
-		imageStore(derivativeMapImage, G_IDX/2, vec4(dx, dy, 0.0, 0.0));
-#elif defined(IRR_DDX_IS_DEFINED)
-	float d = DDX(textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor);
-	if ((G_IDX.x & 1) == 0 && (G_IDX.y & 1) == 0)
-		imageStore(derivativeMapImage, G_IDX/2, vec4(d, 0.0, 0.0, 0.0));
-#else
-	const uint LC_IDX = gl_LocalInvocationIndex;
-	smem[LC_IDX] = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor;
+	vec2 bumpMapSz = vec2(textureSize(bumpMapSampler, 0));
+
+	const ivec2 LC_IDX = ivec2(gl_LocalInvocationID.xy);
+	smem[getAddr(LC_IDX)] = textureLod(bumpMapSampler, vec2(1.0, 1.0)/vec2(RESOLUTION) * vec2(G_IDX), 0.0).r * uHeightScaleFactor;
+	if (LC_IDX.x == 0 || LC_IDX.x == 15)
```

Comment:
don't do the if-statements

Reply:
ok:
1) Your texture coordinate is wrong, you'd need to offset by 0.5 of a pixel
2) You should use texelFetch instead anyway.
3) Best way to do this is like this

```cpp
for (uint vid=gl_LocalInvocationIndex; vid<SMEM_SIZE; vid+=WORK_GROUP_TOTAL_SIZE)
{
   ivec2 texel = gl_WorkGroupID.xy*gl_WorkGroupSize.xy-ivec2(1,1);
   int row = vid/SMEM_DIM_X;
   texel.x += vid-row*SMEM_DIM_X;
   texel.y += row;
   smem[vid] = texelFetch(bumpMapSampler,texel,0).r;
}
```

compiler should unroll the loop and you will just get the texel calculation math + 1 conditional texel fetch

Reply:
hm smart
But `texelFetch` require that bump map and derivative map are the same size + nothing-doing threads if the size is not multiple of 16. Filtering is so costly?

Reply:
bump map and drivative map should absolutely be the same size, because from frequency analysis theory (mixing fourier and Shannon Nyquist):
1. A bump-map of resolution N contains frequencies of at most K Hz
2. A derivative applied before a fourier transform does introduces a mutliplicative term to the original fourier transform, hence if the original did not contain frequencies above K, then the fourier of the derivative will also not contain any frequencies above K
3. Hence if derivative map is lower resolution them it looses detail or if its higher its wasting pixels because the signal can be reconstructed anyway from a lower amount of samples